scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordered_serialization/providers/TraversablesOrderedBuf.scala

```diff
@@ -0,0 +1,242 @@
+        val len = freshT("len")
+        val firstVal = freshT("firstVal")
+        val travBuilder = freshT("travBuilder")
+        val iter = freshT("iter")
+        val extractionTree = maybeArray match {
+          case IsArray =>
+            q"""val $travBuilder = new Array[..$innerTypes]($len)
+            var $iter = 0
+            while($iter < $len) {
+              $travBuilder($iter) = ${innerBuf.get(inputStream)}
+              $iter = $iter + 1
+            }
+            $travBuilder : $outerType
+            """
+          case NotArray =>
+            q"""val $travBuilder = $companionSymbol.newBuilder[..$innerTypes]
+            var $iter = 0
+            while($iter < $len) {
+              $travBuilder += ${innerBuf.get(inputStream)}
+              $iter = $iter + 1
+            }
+            $travBuilder.result : $outerType
+            """
+        }
+        q"""
+        val $len: Int = $inputStream.readSize
+        if($len > 0)
+        {
+          if($len == 1) {
+            val $firstVal: $innerType = ${innerBuf.get(inputStream)}
+            $companionSymbol.apply($firstVal) : $outerType
+          } else {
+            $extractionTree : $outerType
+          }
+        } else {
+          $companionSymbol.empty : $outerType
+        }
+      """
+      }
+
+      override def compare(elementA: ctx.TermName, elementB: ctx.TermName): ctx.Tree = {
+
+        val a = freshT("a")
+        val b = freshT("b")
+        val cmpFnName = freshT("cmpFnName")
+        val innerCmpFn = q"""
+          def $cmpFnName(a: $innerType, b: $innerType): Int = {
+            val $a = a
+            val $b = b
+            ${innerBuf.compare(a, b)}
+          }
+          """
+        maybeSort match {
+          case DoSort =>
+            q"""
+              $innerCmpFn
+              _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.TraversableHelpers.memCompareWithSort($elementA, $elementB)($cmpFnName)
+              """
+
+          case NoSort =>
+            q"""
+              $innerCmpFn
+              _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.TraversableHelpers.memCompare($elementA, $elementB)($cmpFnName)
+              """
+        }
+
+      }
+
+      override val lazyOuterVariables: Map[String, ctx.Tree] = innerBuf.lazyOuterVariables
+
+      override def length(element: Tree): CompileTimeLengthTypes[c.type] = {
+
+        innerBuf.length(q"$element.head") match {
+          case const: ConstantLengthCalculation[_] =>
+            FastLengthCalculation(c)(q"""{
+              sizeBytes($element.size) + $element.size * ${const.toInt}
+            }""")
+          case m: MaybeLengthCalculation[_] =>
+            val maybeRes = freshT("maybeRes")
+            MaybeLengthCalculation(c)(q"""
+              if($element.isEmpty) {
+                _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.DynamicLen(1)
```

Comment:
can you capture this and share with the same below. The 1 here is slightly non-obvious, since it is `sizeBytes(0)`, right?